pkg/controller/directvolumemigration/rsync.go

```diff
@@ -0,0 +1,827 @@
+`
+
+func (t *Task) areRsyncTransferPodsRunning() (bool, error) {
+	// Get client for destination
+	destClient, err := t.getDestinationClient()
+	if err != nil {
+		return false, err
+	}
+
+	pvcMap := t.getPVCNamespaceMap()
+
+	selector := labels.SelectorFromSet(map[string]string{
+		"app":     "directvolumemigration-rsync-transfer",
+		"owner":   "directvolumemigration",
+		"purpose": "rsync",
+	})
+	for ns, _ := range pvcMap {
+		pods := corev1.PodList{}
+		err = destClient.List(
+			context.TODO(),
+			&k8sclient.ListOptions{
+				Namespace:     ns,
+				LabelSelector: selector,
+			},
+			&pods)
+		if err != nil {
+			return false, err
+		}
+		for _, pod := range pods.Items {
+			if pod.Status.Phase != corev1.PodRunning {
+				return false, nil
+			}
+		}
+	}
+
+	return true, nil
+
+	// Create rsync transfer pod on destination
+
+	// Create rsync client pod on source
+}
+
+// Generate SSH keys to be used
+// TODO: Need to determine if this has already been generated and
+// not to regenerate
+func (t *Task) generateSSHKeys() error {
+	// Check if already generated
+	if t.SSHKeys != nil {
+		return nil
+	}
+	// Private Key generation
+	privateKey, err := rsa.GenerateKey(rand.Reader, 4096)
+	if err != nil {
+		return err
+	}
+
+	// Validate Private Key
+	err = privateKey.Validate()
+	if err != nil {
+		return err
+	}
+
+	t.SSHKeys = &sshKeys{
+		PublicKey:  &privateKey.PublicKey,
+		PrivateKey: privateKey,
+	}
+	return nil
+}
+
+func (t *Task) createRsyncConfig() error {
+	// Get client for destination
+	destClient, err := t.getDestinationClient()
+	if err != nil {
+		return err
+	}
+	// Get client for source
+	srcClient, err := t.getSourceClient()
+	if err != nil {
+		return err
+	}
+
+	// Create rsync configmap/secret on source + destination
+	// Create rsync secret (which contains user/pass for rsync transfer pod) in
+	// each namespace being migrated
+	// Needs to go in every namespace where a PVC is being migrated
+	pvcMap := t.getPVCNamespaceMap()
+
+	for ns, vols := range pvcMap {
+		pvcList := []pvc{}
+		for _, vol := range vols {
+			pvcList = append(pvcList, pvc{Name: vol})
+		}
+		// Generate template
+		rsyncConf := rsyncConfig{
+			SshUser:   "root",
+			Namespace: ns,
+			PVCList:   pvcList,
+			Password:  "changeme",
+		}
+		var tpl bytes.Buffer
+		temp, err := template.New("config").Parse(rsyncConfigTemplate)
+		if err != nil {
+			return err
+		}
+		err = temp.Execute(&tpl, rsyncConf)
+		if err != nil {
+			return err
+		}
+
+		configMap := corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: ns,
+				Name:      "directvolumemigration-rsync-config",
+				Labels: map[string]string{
+					"app": "directvolumemigration-rsync-transfer",
+				},
+			},
+		}
+		err = yaml.Unmarshal(tpl.Bytes(), &configMap)
+		if err != nil {
+			return err
+		}
+
+		// Create configmap on source + dest
+		// Note: when this configmap changes the rsync pod
+		// needs to restart
+		// Need to launch new pod when configmap changes
```

Comment:
this is where I think versioning of configs is needed, to understand what running pod is carrying what config

Reply:
Yes either that or regenerate this configmap on every run. Agree this current workflow isn't sustainable